### PR TITLE
YLP-236 Display zen mode

### PR DESCRIPTION
**Pull request by franck-fourel**
- add Zen mode and Confidential mode management in tideline data, deduplicate based on eventID and most recent inputTime
- display the zen mode event in the Glucose and Event section. 
- No tool tip
- Displayed in the background, all other events are on top of it.

### CODE REVIEW

File: packages/tideline/css/tideline-colors.less [hunk 1]
```diff
@@ -71,6 +71,10 @@
 // physical activty colors
 @physicalactivity: #00D3E6;
 
+// Zen mode Event colors
+@zenmode: #9FCC93;
+@zentext: #85B179;
```
Comment (yvon-dblg) [inline packages/tideline/css/tideline-colors.less:76]: I would have set the colors in https://github.com/mdblp/blip/blob/dblp/packages/viz/src/styles/colors.css
And use the var() css function to get it

File: packages/tideline/js/plot/zenModeEvent.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+ * FOR A PARTICULAR PURPOSE. See the License for more details.
+ *
+ * You should have received a copy of the License along with this program; if
+ * not, you can obtain one from Tidepool Project at tidepool.org.
+ * == BSD2 LICENSE ==
+ */
+
+var d3 = require('d3');
+var _ = require('lodash');
+const { duration } = require('moment-timezone');
```
Comment (yvon-dblg) [inline packages/tideline/js/plot/zenModeEvent.js:20]:
```suggestion
```
Import not use.

File: packages/tideline/js/plot/zenModeEvent.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  var xPos = function(d) {
+    return opts.xScale(Date.parse(d.normalTime));
+  };
+  var calculateWidth = function(d) {
+    var s = Date.parse(d.normalTime);
+    var units = d.duration.units;
+    var msfactor = 1000;
+    switch (units) {
+      case 'seconds':
+        msfactor = msfactor;
```
Comment (yvon-dblg) [inline packages/tideline/js/plot/zenModeEvent.js:44]:
```suggestion
```
Not needed

File: packages/tideline/js/plot/zenModeEvent.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        break;
+      case 'minutes': 
+        msfactor = msfactor * 60;
+        break;
+      case 'hours':
+        msfactor = msfactor * 60 * 60;
+        break;
+      default:
+        msfactor = msfactor;
+        break;
```
Comment (yvon-dblg) [inline packages/tideline/js/plot/zenModeEvent.js:54]:
```suggestion
```
Not needed

File: packages/tideline/js/plot/zenModeEvent.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    var e = Date.parse(dt.addDuration(s, d.duration.value * msfactor)); 
+    return opts.xScale(e) - opts.xScale(s);
+  };
+  var height = pool.height() - 20
```
Comment (yvon-dblg) [inline packages/tideline/js/plot/zenModeEvent.js:59]:
```suggestion
  var height = pool.height() - 20;
```

File: packages/tideline/js/plot/zenModeEvent.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+ * not, you can obtain one from Tidepool Project at tidepool.org.
+ * == BSD2 LICENSE ==
+ */
+
+var d3 = require('d3');
+var _ = require('lodash');
+const { duration } = require('moment-timezone');
+var dt = require('../data/util/datetime');
+
+// var drawZenModeEvent = require('./util/drawZenEvent');
```
Comment (yvon-dblg) [inline packages/tideline/js/plot/zenModeEvent.js:23]:
```suggestion
```

File: packages/tideline/test/tidelinedata_test.js [hunk 7]
```diff
@@ -1009,4 +1009,58 @@ describe('TidelineData', function() {
...
+
+    it('should be a function', function() {
+      assert.isFunction(thisTd.setEvents);
+    });
+
+    const Zen1 = _.filter(thisTd.zenEvents, { 'eventId': 'Event1' });
+    const Zen2 = _.filter(thisTd.zenEvents, { 'eventId': 'Event2' });
+    const Zen3 = _.filter(thisTd.zenEvents, { 'eventId': 'Event3' });
+
+    console.log(thisTd.zenEvents);
```
Comment (yvon-dblg) [inline packages/tideline/test/tidelinedata_test.js:1047]:
```suggestion
```

File: packages/tideline/test/tidelinedata_test.js [hunk 8]
```diff
@@ -1009,4 +1009,58 @@ describe('TidelineData', function() {
...
+    it('should deduplicate Events based on eventId', function() {
+      // PA1, PA2 and undefined
```
Comment (yvon-dblg) [inline packages/tideline/test/tidelinedata_test.js:1049]: Should we test a de-duplication ?

**Review by yvon-dblg — CHANGES_REQUESTED** [review]
- A couple of lint cleanup
- I would have use let/const instead of var, but it's not a mandatory change.
- Missing a test case

File: packages/tideline/test/tidelinedata_test.js (hunk 8 above)
Comment (franck-fourel) [reply packages/tideline/test/tidelinedata_test.js]: That's what is done for Event3, 2 events as input and 1 event as output in zenEvents

File: packages/tideline/js/plot/zenModeEvent.js [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+          return opts.r;
+        },
+        'stroke-width': 0,
+        class: 'd3-circle-zen',
+        id: function(d) {
+          return 'zen_' + d.id;
+        }
+      });
+      zenGroup.append('text')
+        .text('ZEN')
```
Comment (yvon-dblg) [inline packages/tideline/js/plot/zenModeEvent.js:109]: Should we add a translation here ?

Comment (franck-fourel) [reply packages/tideline/js/plot/zenModeEvent.js:109]: good point !!!!
@MarineDBL is `ZEN` supposed to be translated at some point ?

Comment (MarineDBL) [reply packages/tideline/js/plot/zenModeEvent.js:109]: Nope

**Review by yvon-dblg — APPROVED** [review]
LGTM

**Pull request merged**